Refactor tasting code and reagent taste carrying, no longer taste your stomach contents or have non-existing flavours win out

00-Steven opened the pull request:
## About The Pull Request

Soooooooooooooooooooooooooooooooo tasting code has been jank for a good damn while, and with this we aim to fix that- let's start from the beginning.

Earlier when I was making a food item, I noticed that it'd never actually have the taste I gave it, just tasting indescribable as things without tastes do.
Weird, cause it definitely has a taste set, and plenty of nutriment subtype reagents!...
When looking into it, I noticed that we only copy the tastes to *specifically* the base nutriment and the vitamin subtype:
https://github.com/tgstation/tgstation/blob/57b718ad20c3b6af02dbdbc87f3cae4dd4d7e297/code/datums/components/food/edible.dm#L215-L216
But! Surely the other reagents should have tastes, right?.... No:
https://github.com/tgstation/tgstation/blob/57b718ad20c3b6af02dbdbc87f3cae4dd4d7e297/code/modules/reagents/chemistry/reagents/food_reagents.dm#L119-L120
we were always returning the tastes data, but never actually setting it except for those two reagents, overriding any and all tastes.
This caused the non-taste of the proteins and fats to override the food taste applied to the vitamins with their volume.

So, to fix this, we replace these hardcoded ids with a check for `carry_food_tastes` on the nutriment reagent, such that each can individually define whether they carry food tastes. This lets oils have their own tastes, while letting all other nutriments carry the tastes.
We then make `get_taste_description(...)` return the base flavour if no special flavour is found, and add base flavours to proteins and vitamins based on what pure proteins/vitamins with no further additives would roughly taste like.

This fixes our issue!....... Except, we now always taste something "rich earthy pungent" when we eat. Looking into it, that's the organ tissue reagent, used in our stomachs:
https://github.com/tgstation/tgstation/blob/57b718ad20c3b6af02dbdbc87f3cae4dd4d7e297/code/modules/reagents/chemistry/reagents/food_reagents.dm#L232-L236
....Because whenever we ingest something, we taste the contents of our stomach. This was previously less noticeable, as while it *did* count, it only added a weight for _no flavour_.

To fix this we just, well, make it so you taste what you're eating and not the container itself.
It's as simple as separating the tasting code from the reagent holder, and making the tasting that happens on ingest exposure use that logic instead.
We keep the old logic for the INTJ skillchip and tasting machine, just re-routed through this, and split up some procs to avoid too much duplication.

This fixes our issue! Though, there's still some minor issues.
This lets us add organ tissue instead of nutriment to other organs so they no longer taste indescribable, as it actually works properly now, but letting organ tissue hold flavours would risk having the stomach organ tissue permanently leech the flavours of the organ tissue you eat.
So we make a subtype of organ tissue for the stomach, stomach lining, which cannot hold flavours.
This lets us apply the base organ tissue to all other organs without risking stomach contamination.
## Why It's Good For The Game

It's jank to taste your stomach contents.
It's jank for organs to taste like nothing.
It's jank for reagents with flavours to never taste like anything.
It's jank for tasteless reagents to block out food flavours.
It's jank for organs to only sometimes contain organ tissues.
yeah.
## Changelog
:cl:
refactor: Refactored tasting code. Please report any jank tasting behaviours.
fix: You taste what you eat instead of the contents of your own stomach.
fix: Other basic nutriment subtypes also carry food flavours, instead of outweighing the food flavours with their lack of flavour.
fix: Basic nutriment subtypes have a flavour they default to, in case of not carrying any food flavours.
fix: Oils taste like oil instead of tasting indescribable.
fix: Clothing eaten by moths actually carries its flavours instead of tasting indescribable.
fix: Organs use organ tissue instead of nutriment.
fix: You can actually taste organ tissue.
qol: Mineral slurry tastes like minerals instead of tasting indescribable.
fix: Tasted flavours can actually be "strong" in addition to mild or "a hint of", as intended.
qol: Flavours are only strong when their percentage of what you're tasting is more than 4x taste sensitivity instead of 3x, for most default tongues being 60% and lizards 40% of the drink.
qol: Tasted flavours are sorted by strength, stronger flavours coming first.
qol: Tasted flavours are grouped into "strong", "mild", and "hint" in the message, replacing "You can taste a hint of bitterness, alcohol, a hint of sweetness, and a hint of cola" with "You can taste alcohol and a hint of bitterness, sweetness, and cola".
/:cl:

00-Steven committed:
make reagent list tasting proc

00-Steven committed:
split up taste()

00-Steven committed:
taste_list not taste

00-Steven committed:
taste_container not taste

00-Steven committed:
move reagent holder taste proc to new system

00-Steven committed:
un-hardcode taste carrying

00-Steven committed:
default to normal taste if not carrying any

00-Steven committed:
add default flavours

00-Steven committed:
oil doesn't carry taste

00-Steven committed:
neither does specifically stomach lining

00-Steven committed:
add organ tissue (finally)

00-Steven committed:
don't include your debug statements like a dumbass

00-Steven committed:
swapsies

MrMelbert commented on code/datums/components/food/edible.dm:
```diff
@@ -212,10 +212,12 @@ Behavior that's still missing from this component that original food items had t
-			owner.reagents.add_reagent(rid, amount, added_purity = reagent_purity)
+		if(length(tastes) && ispath(rid, /datum/reagent/consumable/nutriment))
+			var/datum/reagent/consumable/nutriment/nid = rid
+			if(nid.carry_food_tastes)
```
Should this be `iniital(nid.carry_food_tastes)`? It looks like `rid` is a typepath

00-Steven replied:
works as is, buuuuuuuut ...yeah probably

MrMelbert commented on code/modules/reagents/chemistry/reagents/food_reagents.dm:
```diff
@@ -66,6 +66,9 @@
 	color = "#664330" // rgb: 102, 67, 48
 	chemical_flags = REAGENT_CAN_BE_SYNTHESIZED
 
+	// Whether this reagent should get the tastes of food it's in applied onto it
+	var/carry_food_tastes = TRUE
```
```suggestion
	/// Whether this reagent should get the tastes of food it's in applied onto it
	var/carry_food_tastes = TRUE
```

MrMelbert commented:
Also I don't really know if it's even worth adding organ tissue to a bunch more organs? 

It was specifically added (#55966) to fix (in a kind-of-jank way) the fact that the "nutriments" in a "stomach" would merge with nutriments you eat, thus gaining their tastes

00-Steven commented:
> Also I don't really know if it's even worth adding organ tissue to a bunch more organs?
> 
> It was specifically added (https://github.com/tgstation/tgstation/pull/55966) to fix (in a kind-of-jank way) the fact that the "nutriments" in a "stomach" would merge with nutriments you eat, thus gaining their tastes

Ah, yeah no this doesn't change that logic, it's just shifting the reagents each use now that the system's more flexible so organs can have saner tastes.
Regular organs use organ tissue, stomachs use the organ tissue subtype "Stomach Lining" with `carry_food_tastes = FALSE`.

00-Steven committed:
Update code/modules/reagents/chemistry/reagents/food_reagents.dm

Co-authored-by: MrMelbert <51863163+MrMelbert@users.noreply.github.com>

00-Steven committed:
initial that mfer

00-Steven committed:
Merge branch 'hmmmtodaywerefactortastingactually' of https://github.com/00-Steven/tgstation into hmmmtodaywerefactortastingactually

00-Steven commented:
It does remove a bunch of the ways in which that was a jank fix
You no longer taste the contents of your stomach, for one

Though ideally I'd want stomachs to be able to have their own flavours without them leeching, doing that's still a whole mess of its own

SmArtKar commented on code/modules/reagents/chemistry/taste.dm:
```diff
@@ -0,0 +1,44 @@
+				if(percent < minimum_percent)
+					continue
+				var/intensity_desc = "a hint of"
+				if(percent > minimum_percent * 2 || percent == 100)
+					intensity_desc = ""
+				else if(percent > minimum_percent * 3)
+					intensity_desc = "the strong flavor of"
```
Else if never triggers here, percent cannot be more than 3 times than minimum_percent but less than two times at the same time

00-Steven replied:
lmfao, didn't notice this when copying it over from the old code
good catch, thanks o7

00-Steven replied:
done
![image](https://github.com/user-attachments/assets/d4876147-efe1-4ba2-98cc-701896c9f7f5)
...but actually lemme maybe just like, redo that. entirely. it results in gross sentences like
![image](https://github.com/user-attachments/assets/0ec56f7a-8cd0-416a-87cf-4340b5fca805)

00-Steven committed:
rewrite the rest of the horse

00-Steven commented:
okay so! in that last commit I rewrote the entire goddamn proc, to avoid situations like:
> "You can taste the strong flavor of cola, a hint of bitterness, and the strong flavor of sweetness."

Which is both stupid and a mouthful.
So I've addressed the TODO left in this code \**checks notes*\* on Feb 21st 2017, about seven and a half years ago.
This TODO has been here for seven and a half years.
This broken if statement has been here for seven and a half years, since implementation.
Man.

Actually some of the other code I've addressed in this before is apparently similarly ancient, but hey-

Anyhow now it sorts it by taste as said in the TODO, but also groups these into 'strong', 'mild', and 'hint' tastes!
This avoids the previous, and instead makes it:
> "You can taste the strong flavor of cola and sweetness and a hint of bitterness."
> "You can taste bitterness and cola."
> "You can taste the strong flavor of cola, some bitterness, and a hint of sweetness"

It also no longer counts up the list after the fact for total taste, and just keeps a running total as it goes.
Finally I've increased the multiplier for strong from 3 to 4, such that at least for the default taste sensitivity we need at least 60% of the taste before being considered "strong".
I still dislike how this says "the strong flavor of" for single reagent drinks, but it being mild until mixed is also off, while nuking percentual taste for absolute taste or some other system would be a whole thing on its own requiring far more thought about our individual taste strengths than this.

00-Steven commented:
aight so, I've settled on "as the original dev intended it, until I come up with anything better"

00-Steven committed:
one taste only mild

00-Steven commented:
groaaan conflicts

00-Steven committed:
Merge branch 'master' into hmmmtodaywerefactortastingactually
